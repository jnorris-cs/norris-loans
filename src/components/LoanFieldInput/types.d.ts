import type { InputValue } from "../../types";

export type FieldChangeAction = Pick<State, "hasError" | "errorMessage"> & {
	value: InputValue;
};

export type StandardInputProps = {
	onChange: (payload: FieldChangeAction) => void;
	onBlur: () => void;
	onFocus: () => void;
	name: string;
	id: string;
};

export type DateLoanFieldInputProps = {
	field: FieldMetadata;
	value: string;
} & StandardInputProps;

export type MoneyLoanFieldInputProps = {
	field: MoneyFieldMetadata;
	value?: number;
} & StandardInputProps;

export type StringLoanFieldInputProps = {
	field: StringFieldMetadata;
	value: string;
} & StandardInputProps;
