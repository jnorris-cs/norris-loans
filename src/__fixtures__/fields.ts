import type { FieldMetadata } from '../types';

import fields from '../../public/json/fieldsMetaData.json';

export default fields as FieldMetadata[];
