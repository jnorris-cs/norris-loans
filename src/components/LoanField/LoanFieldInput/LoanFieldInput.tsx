import type { FieldMetadata, StringFieldMetadata } from "../../../types";

import LoanStringField from "../LoanStringField/LoanStringField";

// TODO move to utils
const isStringField = (field: FieldMetadata): field is StringFieldMetadata =>  {
  return field.type === 'string'
}

interface  LoanFieldInputProps {
	field: FieldMetadata;
	value?: unknown;
}

const LoanFieldInput = ({field}: LoanFieldInputProps) => {

  if (isStringField(field)){
    return <LoanStringField field={field} />
  }

  return false
};


export default LoanFieldInput