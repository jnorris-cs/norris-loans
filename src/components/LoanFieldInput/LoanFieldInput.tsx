import StringLoanFieldInput from './StringLoanFieldInput/StringLoanFieldInput';
import MoneyLoanFieldInput from './MoneyLoanFieldInput/MoneyLoanFieldInput';
import DateLoanFieldInput from './DateLoanFieldInput/DateLoanFieldInput';

import type {
	StandardInputProps,
	DateLoanFieldInputProps,
	MoneyLoanFieldInputProps,
	StringLoanFieldInputProps,
} from './types';
import type { FieldMetadata, InputValue } from '../../types';

const isStringField = (
	props: StandardInputProps
): props is StringLoanFieldInputProps => {
	return props.field.type === 'string';
};

const isDateField = (
	props: StandardInputProps
): props is DateLoanFieldInputProps => {
	return props.field.type === 'date';
};

const isMoneyField = (
	props: StandardInputProps
): props is MoneyLoanFieldInputProps => {
	return props.field.type === 'money';
};

type LoanFieldInputProps = {
	field: FieldMetadata;
	value?: InputValue;
} & StandardInputProps;

const LoanFieldInput = (props: LoanFieldInputProps) => {
	if (isStringField(props)) {
		return <StringLoanFieldInput {...props} />;
	}
	if (isMoneyField(props)) {
		return <MoneyLoanFieldInput {...props} />;
	}
	if (isDateField(props)) {
		return <DateLoanFieldInput {...props} />;
	}

	return false;
};

export default LoanFieldInput;
