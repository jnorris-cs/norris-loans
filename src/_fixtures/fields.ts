import fields from "../../public/json/fieldsMetaData.json";
import type { FieldMetadata } from "../types";

export default fields as FieldMetadata[];
