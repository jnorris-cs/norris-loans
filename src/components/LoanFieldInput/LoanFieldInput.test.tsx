import type { FieldMetadata } from 'types';

import { render } from '@testing-library/react';

import fields from '../../__tests__/fixtures/fields';
import LoanFieldInput from './LoanFieldInput';

const dateField = fields.find((f) => f.type === 'date') as FieldMetadata;
const moneyField = fields.find((f) => f.type === 'string') as FieldMetadata;
const stringField = fields.find((f) => f.type === 'money') as FieldMetadata;

const generatePropsFromField = (field: FieldMetadata) => {
	return {
		field: field,
		id: 'test',
		name: 'test',
		onBlur: () => {},
		onChange: () => {},
		onFocus: () => {},
		value: 'value',
	};
};

describe('LoanFieldInput', () => {
	it('renders date field correct', () => {
		const result = render(
			<LoanFieldInput {...generatePropsFromField(dateField)} />
		);
		expect(result.baseElement.firstChild).toMatchSnapshot();
	});

	it('renders money field correct', () => {
		const result = render(
			<LoanFieldInput {...generatePropsFromField(moneyField)} />
		);
		expect(result.baseElement.firstChild).toMatchSnapshot();
	});

	it('renders string field correct', () => {
		const result = render(
			<LoanFieldInput {...generatePropsFromField(stringField)} />
		);
		expect(result.baseElement.firstChild).toMatchSnapshot();
	});
});
