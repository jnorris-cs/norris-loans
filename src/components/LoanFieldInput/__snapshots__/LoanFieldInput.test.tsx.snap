// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoanFieldInput renders date field correct 1`] = `
<div>
  <input
    field="[object Object]"
    id="test-id"
    name="test-name"
    type="date"
    value="value"
  />
</div>
`;

exports[`LoanFieldInput renders money field correct 1`] = `
<div>
  <input
    id="test-id"
    name="test-name"
    type="text"
    value="value"
  />
</div>
`;

exports[`LoanFieldInput renders string field correct 1`] = `
<div>
  <input
    id="test-id"
    inputmode="decimal"
    name="test-name"
    type="text"
    value="$NaN"
  />
</div>
`;
